File: Sources/Wrap/HgWrap.swift
import Foundation

public class HgWrap: VcsWrap {
  public override func setupDirectory(path: String, packagefilesPath: String) throws {
    try self.assertRequired("hg")
    guard let urlAsString = self.url else { throw WrapError.genericError("Expected URL to clone") }
    if let url = URL(string: urlAsString) {
      guard let rev = self.revision else { throw WrapError.genericError("Missing revision") }
      // Do something like https://github.com/mesonbuild/meson/blob/3e7c08f358e9bd91808c8ff3b76c11aedeb82f85/mesonbuild/wrap/wrap.py#L549
      let targetDirectory =
        self.directory ?? url.lastPathComponent.replacingOccurrences(of: ".git", with: "")
      let fullPath = path + "/" + targetDirectory
      try self.executeCommand(["hg", "clone", url.description, fullPath])
      if rev.lowercased() != "tip" {
        try self.executeCommand(["hg", "--cwd", fullPath, "checkout", rev])
      }
      try self.postSetup(path: fullPath, packagesfilesPath: packagefilesPath)
    } else {
      throw WrapError.genericError("Malformed URL: \(String(describing: self.url))")
    }
  }
}

File: src/libwrap/hgwrap.cpp
#include "log.hpp"
#include "polyfill.hpp"
#include "utils.hpp"
#include "wrap.hpp"

#include <cctype>
#include <filesystem>
#include <string>
#include <vector>

const static Logger LOG("wrap::HgWrap"); // NOLINT

bool HgWrap::setupDirectory(const std::filesystem::path &path,
                            const std::filesystem::path &packageFilesPath) {
  auto url = this->url;
  if (url.empty()) {
    this->errors.emplace_back("Missing URL");
    LOG.warn("URL is empty");
    return false;
  }
  if (this->revision.empty()) {
    this->errors.emplace_back("Missing revision");
    LOG.warn("Revision is empty");
    return false;
  }
  std::string rev = this->revision;
  if (this->directory->empty()) {
    this->errors.emplace_back("Missing directory");
    LOG.warn("Directory is empty");
    return false;
  }
  auto targetDirectory = this->directory.value();
  std::string const fullPath =
      std::format("{}/{}", path.generic_string(), targetDirectory);
  auto result =
      launchProcess("hg", std::vector<std::string>{"clone", url, fullPath});
  if (result) {
    auto isTip = rev.size() == 3 &&
                 (std::tolower(rev[0]) == 't' && std::tolower(rev[1]) == 'i' &&
                  std::tolower(rev[2]) == 'p');
    if (isTip) {
      result = launchProcess(
          "hg", std::vector<std::string>{"--cwd", fullPath, "checkout", rev});
      if (!result) {
        this->errors.emplace_back("Failed to do hg checkout");
        return false;
      }
    }
    return this->postSetup(fullPath, packageFilesPath);
  }
  this->errors.emplace_back("Failed to do hg clone");
  return false;
}
